core/lib/zksync_core/src/state_keeper/keeper.rs

```diff
@@ -274,23 +274,50 @@ impl ZkSyncStateKeeper {
             .with_context(|| format!("failed loading upgrade transaction for {protocol_version:?}"))
     }
 
-    async fn wait_for_new_batch_env(
+    async fn wait_for_new_batch_params(
         &mut self,
         cursor: &IoCursor,
-    ) -> Result<(SystemEnv, L1BatchEnv), Error> {
+    ) -> Result<L1BatchParams, Error> {
         while !self.is_canceled() {
-            if let Some(envs) = self
+            if let Some(params) = self
                 .io
-                .wait_for_new_batch_env(cursor, POLL_WAIT_DURATION)
-                .await
-                .context("error waiting for new L1 batch environment")?
+                .wait_for_new_batch_params(cursor, POLL_WAIT_DURATION)
+                .await?
             {
-                return Ok(envs);
+                return Ok(params);
             }
         }
         Err(Error::Canceled)
     }
 
+    async fn wait_for_new_batch_env(
```

Comment:
This fixes state keeper hanging up during node shutdown if the requested L1 batch root hash is not being computed. It was technically an issue even without introducing tree data fetcher; I've just encountered this problem during local testing and decided it makes sense to fix.